chainer/training/updaters/multiprocess_parallel_updater.py

```diff
@@ -0,0 +1,376 @@
+        if self._initialized:
+            return
+        self._initialized = True
+
+        self._master.cleargrads()
+        for i in six.moves.range(1, len(self._devices)):
+            pipe, worker_end = multiprocessing.Pipe()
+            worker = _Worker(i, worker_end, self)
+            worker.start()
+            self._workers.append(worker)
+            self._pipes.append(pipe)
+
+        with cuda.Device(self._devices[0]):
+            self._master.to_gpu(self._devices[0])
+            if len(self._devices) > 1:
+                comm_id = nccl.get_unique_id()
+                self._send_message(("set comm_di", comm_id))
+                self.comm = nccl.NcclCommunicator(len(self._devices),
+                                                  comm_id, 0)
+
+    def update_core(self):
+        self.setup_workers()
+
+        self._send_message(('update', None))
+        with cuda.Device(self._devices[0]):
+            # For reducing memory
+            self._master.cleargrads()
+
+            optimizer = self.get_optimizer('main')
+            batch = self.get_iterator('main').next()
+            batch = self.converter(batch, self._devices[0])
+
+            loss = _calc_loss(self._master, batch)
+
+            self._master.cleargrads()
+            loss.backward()
+
+            # NCCL: reduce grads
+            null_stream = cuda.Stream.null
+            if self.comm is not None:
+                gg = gather_grads(self._master)
+                self.comm.reduce(gg.data.ptr, gg.data.ptr, gg.size,
+                                 nccl.NCCL_FLOAT, nccl.NCCL_SUM,
+                                 0, null_stream.ptr)
+                scatter_grads(self._master, gg)
+                del gg
+            optimizer.update()
+            if self.comm is not None:
+                gp = gather_params(self._master)
+                self.comm.bcast(gp.data.ptr, gp.size, nccl.NCCL_FLOAT,
+                                0, null_stream.ptr)
+
+    def finalize(self):
+        self._send_message(('finalize', None))
+
+        for worker in self._workers:
+            print("join", worker)
+            worker.join()
+
+
+def _calc_loss(model, in_arrays):
+    if isinstance(in_arrays, tuple):
+        in_vars = tuple(variable.Variable(x) for x in in_arrays)
+        return model(*in_vars)
+    elif isinstance(in_arrays, dict):
+        in_vars = {key: variable.Variable(x)
+                   for key, x in six.iteritems(in_arrays)}
+        return model(**in_vars)
+    else:
+        in_vars = variable.Variable(in_arrays)
+        return model(in_vars)
+
+
+def size_num_grads(link):
+    """Count total size of all gradient arrays of a given link
+
+    Args:
+        link (chainer.link.Link): Target link object.
+    """
+    size = 0
+    num = 0
+    for param in link.params():
+        if param.size == 0:
+            continue
+        size += param.size
+        num += 1
+    return size, num
+
+
+_batch_memcpy = cuda.cupy.ElementwiseKernel(
```

Comment:
It is possible that `cupy` is unavailable. This statement should be executed only when `_available` is True. Is this correct?